File: seqspec/seqspec_methods.py
"""Methods module for seqspec.

This module provides functionality to convert seqspec files into methods sections.
"""
from pathlib import Path
from argparse import ArgumentParser, RawTextHelpFormatter, Namespace

from seqspec.utils import load_spec
from seqspec.Assay import Assay
from seqspec.Region import Region
from seqspec.Read import Read, File


def setup_methods_args(parser) -> ArgumentParser:
    """Create and configure the methods command subparser."""
    subparser = parser.add_parser(
        "methods",
        description="""
Convert seqspec file into methods section.

Examples:
seqspec methods -m rna -o methods.txt spec.yaml  # Save methods section to file
seqspec methods -m rna spec.yaml                 # Print methods section to stdout
---
""",
        help="Convert seqspec file into methods section",
        formatter_class=RawTextHelpFormatter,
    )
    subparser_required = subparser.add_argument_group("required arguments")

    subparser.add_argument("yaml", help="Sequencing specification yaml file", type=str)
    subparser_required.add_argument(
        "-m",
        "--modality",
        metavar="MODALITY",
        help="Modality",
        type=str,
        required=True,
    )
    subparser.add_argument(
        "-o",
        "--output",
        metavar="OUT",
        help="Path to output file",
        type=Path,
        default=None,
    )
    return subparser


def validate_methods_args(parser: ArgumentParser, args: Namespace) -> None:
    """Validate the methods command arguments."""
    if not Path(args.yaml).exists():
        parser.error(f"Input file does not exist: {args.yaml}")

    if args.output and args.output.exists() and not args.output.is_file():
        parser.error(f"Output path exists but is not a file: {args.output}")


def run_methods(parser: ArgumentParser, args: Namespace) -> None:
    """Run the methods command."""
    validate_methods_args(parser, args)

    spec = load_spec(args.yaml)
    methods_text = methods(spec, args.modality)

    if args.output:
        args.output.write_text(methods_text)
    else:
        print(methods_text)


def methods(spec: Assay, modality: str) -> str:
    """Generate methods section for spec and modality."""
    m = f"""Methods
The {modality} portion of the {spec.name} assay was generated on {spec.date}.
    """
    m += format_library_spec(spec, modality)
    return m


# TODO: manage sequence/library protocol/kit for cases where each modality has different protocols/kits
def format_library_spec(spec: Assay, modality: str) -> str:
    """Format library specification for methods section."""
    leaves = spec.get_libspec(modality).get_leaves()

    lib_prot = None
    if isinstance(spec.library_protocol, str):
        lib_prot = spec.library_protocol
    elif isinstance(spec.library_protocol, list):
        for i in spec.library_protocol:
            if i.modality == modality:
                lib_prot = i.protocol_id

    lib_kit = None
    if isinstance(spec.library_kit, str):
        lib_kit = spec.library_kit
    elif isinstance(spec.library_kit, list):
        for i in spec.library_kit:
            if i.modality == modality:
                lib_kit = i.kit_id

    seq_prot = None
    if isinstance(spec.sequence_protocol, str):
        seq_prot = spec.sequence_protocol
    elif isinstance(spec.sequence_protocol, list):
        for i in spec.sequence_protocol:
            if i.modality == modality:
                seq_prot = i.protocol_id

    seq_kit = None
    if isinstance(spec.sequence_kit, str):
        seq_kit = spec.sequence_kit
    elif isinstance(spec.sequence_kit, list):
        for i in spec.sequence_kit:
            if i.modality == modality:
                seq_kit = i.kit_id

    s = f"""
Libary structure\n
The library was generated using the {lib_prot} library protocol and {lib_kit} library kit. The library contains the following elements:\n
"""
    for idx, r in enumerate(leaves, 1):
        s += format_region(r, idx)
    s += f"""
\nSequence structure\n
The library was sequenced on a {seq_prot} using the {seq_kit} sequencing kit. The library was sequenced using the following configuration:\n
"""
    reads = spec.get_seqspec(modality)
    for idx, r in enumerate(reads, 1):
        s += format_read(r, idx)
    return s


def format_region(region: Region, idx: int = 1) -> str:
    """Format region for methods section."""
    s = f"{idx}. {region.name}: {region.min_len}-{region.max_len}bp {region.sequence_type} sequence ({region.sequence})"
    if region.onlist:
        s += f", onlist file: {region.onlist.filename}.\n"
    else:
        s += ".\n"
    return s


def format_read(read: Read, idx: int = 1) -> str:
    """Format read for methods section."""
    s = f"- {read.name}: {read.max_len} cycles on the {'positive' if read.strand == 'pos' else 'negative'} strand using the {read.primer_id} primer. The following files contain the sequences in Read {idx}:\n"
    if read.files:
        for idx, f in enumerate(read.files, 1):
            s += "  " + format_read_file(f, idx)
    return s


def format_read_file(file: File, idx: int = 1) -> str:
    """Format read file for methods section."""
    return f"- File {idx}: {file.filename}\n"

File: src/seqspec_methods.rs
use crate::models::assay::Assay;
use crate::models::file::File as ReadFile;
use crate::models::read::Read;
use crate::models::region::Region;
use crate::utils;
use clap::Args;
use std::fs;
use std::io::Write;
use std::path::PathBuf;

#[derive(Debug, Args)]
pub struct MethodsArgs {
    #[clap(help = "Sequencing specification yaml file", required = true)]
    yaml: PathBuf,

    #[clap(
        short,
        long,
        help = "Modality",
        value_name = "MODALITY",
        required = true
    )]
    modality: String,

    #[clap(short, long, help = "Path to output file", value_name = "OUT")]
    output: Option<PathBuf>,
}

pub fn run_methods(args: &MethodsArgs) {
    validate_methods_args(args);
    let spec = utils::load_spec(&args.yaml);

    let text = seqspec_methods(&spec, &args.modality);
    if let Some(out) = &args.output {
        let mut f = fs::File::create(out).unwrap();
        write!(f, "{}", text).unwrap();
    } else {
        println!("{}", text);
    }
}

fn validate_methods_args(args: &MethodsArgs) {
    if !args.yaml.exists() {
        eprintln!("Please use `seqspec methods -h` for help.");
        std::process::exit(1);
    }
    if let Some(out) = &args.output {
        if out.exists() && !out.is_file() {
            eprintln!("Output path exists but is not a file: {}", out.display());
            std::process::exit(1);
        }
    }
}

pub fn seqspec_methods(spec: &Assay, modality: &str) -> String {
    let mut m = format!(
        "Methods\nThe {} portion of the {} assay was generated on {}.\n    ",
        modality, spec.name, spec.date
    );
    m.push_str(&format_library_spec(spec, modality));
    m
}

fn format_library_spec(spec: &Assay, modality: &str) -> String {
    let leaves = spec
        .get_libspec(modality)
        .expect("modality not found")
        .get_leaves();

    let lib_prot: Option<String> = match &spec.library_protocol {
        Some(v) => v
            .iter()
            .find(|p| p.modality == modality)
            .map(|p| p.protocol_id.clone()),
        None => None,
    };
    let lib_kit: Option<String> = match &spec.library_kit {
        Some(v) => v
            .iter()
            .find(|k| k.modality == modality)
            .map(|k| k.kit_id.clone()),
        None => None,
    };
    let seq_prot: Option<String> = match &spec.sequence_protocol {
        Some(v) => v
            .iter()
            .find(|p| p.modality == modality)
            .map(|p| p.protocol_id.clone()),
        None => None,
    };
    let seq_kit: Option<String> = match &spec.sequence_kit {
        Some(v) => v
            .iter()
            .find(|k| k.modality == modality)
            .map(|k| k.kit_id.clone()),
        None => None,
    };

    let mut s = String::new();
    s.push_str("\nLibary structure\n\n");
    s.push_str(&format!(
        "The library was generated using the {} library protocol and {} library kit. The library contains the following elements:\n\n",
        lib_prot.unwrap_or_else(|| "None".to_string()),
        lib_kit.unwrap_or_else(|| "None".to_string())
    ));
    for (idx, r) in leaves.iter().enumerate() {
        s.push_str(&format_region(r, (idx + 1) as i32));
    }
    s.push_str("\nSequence structure\n\n");
    s.push_str(&format!(
        "The library was sequenced on a {} using the {} sequencing kit. The library was sequenced using the following configuration:\n\n",
        seq_prot.unwrap_or_else(|| "None".to_string()),
        seq_kit.unwrap_or_else(|| "None".to_string())
    ));
    let reads = spec.get_seqspec(modality);
    for (idx, r) in reads.iter().enumerate() {
        s.push_str(&format_read(r, (idx + 1) as i32));
    }
    s
}

fn format_region(region: &Region, idx: i32) -> String {
    let mut s = format!(
        "{}. {}: {}-{}bp {} sequence ({})",
        idx, region.name, region.min_len, region.max_len, region.sequence_type, region.sequence
    );
    if let Some(ol) = &region.onlist {
        s.push_str(&format!(", onlist file: {}.\n", ol.filename));
    } else {
        s.push_str(".\n");
    }
    s
}

fn format_read(read: &Read, idx: i32) -> String {
    let strand = if read.strand == "pos" {
        "positive"
    } else {
        "negative"
    };
    let mut s = format!(
        "- {}: {} cycles on the {} strand using the {} primer. The following files contain the sequences in Read {}:\n",
        read.name, read.max_len, strand, read.primer_id, idx
    );
    if !read.files.is_empty() {
        for (i, f) in read.files.iter().enumerate() {
            s.push_str(&format_read_file(f, (i + 1) as i32));
        }
    }
    s
}

fn format_read_file(file: &ReadFile, idx: i32) -> String {
    format!("- File {}: {}\n", idx, file.filename)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::load_spec;
    use std::path::PathBuf;

    fn dogma_spec() -> Assay {
        load_spec(&PathBuf::from("tests/fixtures/spec.yaml"))
    }

    #[test]
    fn test_methods_output_rna() {
        let spec = dogma_spec();
        let text = seqspec_methods(&spec, "rna");
        assert!(text.starts_with("Methods\nThe rna portion of the DOGMAseq-DIG/Illumina assay"));
        assert!(text.contains("Libary structure"));
        assert!(text.contains("Sequence structure"));
        // Should mention the 5 RNA leaf regions
        assert!(text.contains("Cell Barcode"));
        assert!(text.contains("umi"));
        assert!(text.contains("cdna"));
    }

    #[test]
    fn test_format_region() {
        let region = Region::new(
            "bc".into(),
            "barcode".into(),
            "Cell Barcode".into(),
            "onlist".into(),
            "NNNNNNNNNNNNNNNN".into(),
            16,
            16,
            None,
            vec![],
        );
        let s = format_region(&region, 1);
        assert_eq!(
            s,
            "1. Cell Barcode: 16-16bp onlist sequence (NNNNNNNNNNNNNNNN).\n"
        );
    }

    #[test]
    fn test_format_region_with_onlist() {
        let onlist = crate::models::onlist::Onlist::new(
            "ol".into(),
            "barcodes.txt".into(),
            "txt".into(),
            0,
            "".into(),
            "local".into(),
            "".into(),
        );
        let region = Region::new(
            "bc".into(),
            "barcode".into(),
            "Cell Barcode".into(),
            "onlist".into(),
            "N".repeat(16),
            16,
            16,
            Some(onlist),
            vec![],
        );
        let s = format_region(&region, 1);
        assert_eq!(s, "1. Cell Barcode: 16-16bp onlist sequence (NNNNNNNNNNNNNNNN), onlist file: barcodes.txt.\n");
    }

    #[test]
    fn test_format_read() {
        let read = Read::new(
            "R1".into(),
            "Read 1".into(),
            "rna".into(),
            "truseq_read1".into(),
            28,
            28,
            "pos".into(),
            vec![],
        );
        let s = format_read(&read, 1);
        assert_eq!(s, "- Read 1: 28 cycles on the positive strand using the truseq_read1 primer. The following files contain the sequences in Read 1:\n");
    }

    #[test]
    fn test_format_read_file() {
        let f = ReadFile::new(
            "f1".into(),
            "reads_R1.fastq.gz".into(),
            "fastq".into(),
            1024,
            "reads_R1.fastq.gz".into(),
            "local".into(),
            "".into(),
        );
        let s = format_read_file(&f, 1);
        assert_eq!(s, "- File 1: reads_R1.fastq.gz\n");
    }

    #[test]
    fn test_methods_other_modalities() {
        let spec = dogma_spec();
        for modality in ["atac", "protein", "tag"] {
            let text = seqspec_methods(&spec, modality);
            assert!(text.contains("Methods"));
            assert!(text.contains(modality));
        }
    }

    #[test]
    fn test_format_library_spec_sections() {
        let spec = dogma_spec();
        let text = format_library_spec(&spec, "rna");
        assert!(text.contains("Libary structure"));
        assert!(text.contains("Sequence structure"));
        assert!(text.contains("library protocol"));
        assert!(text.contains("sequencing kit"));
    }
}
